library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java

```diff
@@ -442,8 +442,7 @@ private Slice readOrSkip(boolean skip) throws IOException
           inputStream = getInputStream(currentPointer);
         }
 
-        if (inputStream != null && currentPointer.offset <
-            fileSystemWAL.fileContext.getFileStatus(currentOpenPath).getLen()) {
+        if (inputStream != null && inputStream.available() != 0) {
```

Comment:
Please check the documentation for available. It can return 0 even when there is more data to read (blocking). I'm sure you have tested this behavior with HDFS, but maybe it will be necessary to actually try to read from the stream till you get an EOF to be sure that there is indeed no more data to read.